atlp-rwanda/dynamites-ecomm-be
# Payment processing with stripe

### Title: Secure Payment Processing with Stripe API

**Description:**

This PR introduces functionality to securely process payments through the Stripe API. Upon successful payment, the `paid` field in the database will be set to `True`. It also handles cases where the payment request is invalid or authentication fails.

**Related User Story:** As a buyer, I want to securely complete my purchase by processing the payment through the Stripe API so that I can receive confirmation of my order and ensure a smooth transaction.

**Acceptance Criteria:**

*   **GIVEN** a valid payment request is sent to the backend
*   **WHEN** the request is authenticated and the request body is valid
    *   [ ] The payment should be successfully processed using the Stripe API
    *   [ ] The order status should be updated to "Paid" in the database
    *   [ ] A confirmation response should be sent to the frontend, including the order number, total cost, and expected delivery date
*   **WHEN** the request body is invalid or authentication fails
    *   [ ] An appropriate error response should be sent to the frontend, describing the validation error
**Technical Requirements:**

*   \[ \] All inputs should be validated for sanity and security
*   \[ \] The Stripe API key should be kept secure and not exposed in the backend code
*   \[ \] The Stripe library should be used to make the payment, charge the customer, and verify the transaction
*   \[ \] The order status should be updated in real-time in both the database and the frontend to reflect the payment status

**Changes Introduced:**

1.  **Payment Processing:**
    
    *   Integrated Stripe API to handle payment processing.
    *   Created service to interact with Stripe and handle payment logic.
2.  **Order Status Update:**
    
    *   Updated order status to "Paid" in the database upon successful payment.
    *   Ensured real-time updates to the frontend regarding payment status.
3.  **Response Handling:**
    
    *   Sent confirmation response to the frontend including order number, total cost, and expected delivery date.
    *   Handled error responses for invalid requests or authentication failures.
4.  **Security and Validation:**
    
    *   Validated all inputs for sanity and security.
    *   Ensured the Stripe API key is securely managed and not exposed in the codebase.

**Files Modified:**

*   `controllers/buyerController.ts`
*   `docs/buyerDocs.ts`
*   `models/orderEntity.ts`
*   `routes/buyerRoutes`

**How to Test:**

1.  **Valid Payment Request:**
    
    *   Send a valid payment request to the backend.
    *   Verify that the payment is processed successfully.
    *   Check that Paid field for that order is set to true in the database.

2.  **Invalid Payment Request:**
    *   Send an invalid payment request to the backend.
    *   Verify that an appropriate error response is sent to the frontend.
3.  **Security Checks:**
    
    *   Ensure that no sensitive information, such as the Stripe API key, is exposed in the codebase.
    *   Validate that all inputs are properly sanitized and secured.

**Screenshots:**



**Additional Notes:**

-   Ensure the Stripe API key is set in the environment variables before testing.
-  Run the tests using `npm run test` to validate the new changes.

**Reviewer Checklist:**

 - [ ]  Code is clean, well-documented, and follows best practices.
 - [ ] Tests have been added and/or updated to cover the new functionality.
 - [ ] The application runs without errors and all functionalities are verified.

* * *

Please review the changes and let me know if there are any issues or further improvements needed. Thank you!

#81
